File: src/main/java/com/cjw/reactivecommunityproject/common/security/config/SecurityConfig.java
package com.cjw.reactivecommunityproject.common.security.config;

import com.cjw.reactivecommunityproject.common.security.filter.jwt.JwtAuthenticationFilter;
import com.cjw.reactivecommunityproject.common.security.filter.jwt.JwtAuthorizationFilter;
import com.cjw.reactivecommunityproject.common.security.handler.JwtAuthenticationFailedHandler;
import com.cjw.reactivecommunityproject.common.security.handler.JwtAuthorizationDeniedHandler;
import com.cjw.reactivecommunityproject.common.security.service.JwtService;
import com.cjw.reactivecommunityproject.server.cache.custom.service.CacheCustomService;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfigurationSource;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {
    private final JwtService jwtService;
    private final CorsConfigurationSource corsConfigurationSource;
    private final CacheCustomService cacheCustomService;

    private static final String[] ALLOW_PATH_URL_LIST = {"/rest/auth/**"};

    private static final String[] PUBLIC_PATH_LIST = {"/webjars/**", "/opendoc.html", "/rest/auth/**", "/rest/test4"};

    @Bean
    public WebSecurityCustomizer webSecurityCustomizer() {
        return web -> web
                .ignoring()
                .requestMatchers(PUBLIC_PATH_LIST);
    }

    @Bean
    public SecurityFilterChain restFilterChain(HttpSecurity http) throws Exception {
        return http
                .logout(AbstractHttpConfigurer::disable)
                .csrf(AbstractHttpConfigurer::disable)
                .formLogin(AbstractHttpConfigurer::disable)
                .headers(AbstractHttpConfigurer::disable)
                .httpBasic(AbstractHttpConfigurer::disable)
                .sessionManagement(AbstractHttpConfigurer::disable)

                .cors(cors -> cors.configurationSource(corsConfigurationSource))

                .addFilterBefore(new JwtAuthenticationFilter(jwtService), UsernamePasswordAuthenticationFilter.class)
                .securityMatcher("/rest/**")
                .authorizeHttpRequests(authorizeRequests ->
                        authorizeRequests
                                .requestMatchers(ALLOW_PATH_URL_LIST)
                                .permitAll()
                                .anyRequest()
                                .access(new JwtAuthorizationFilter(cacheCustomService))
                )

                .exceptionHandling(exceptionHandle ->
                        exceptionHandle
                                .authenticationEntryPoint(new JwtAuthenticationFailedHandler())
                                .accessDeniedHandler(new JwtAuthorizationDeniedHandler())
                )
                .build();
    }
}

File: src/main/kotlin/com/devcjw/reactivecommunity/auth/config/SecurityConfig.kt
package com.devcjw.reactivecommunity.auth.config

import com.devcjw.reactivecommunity.auth.config.handler.RcJwtFilter
import com.devcjw.reactivecommunity.auth.service.JwtService
import lombok.RequiredArgsConstructor
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authorization.ReactiveAuthorizationManager
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.SecurityWebFiltersOrder
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.web.server.SecurityWebFilterChain
import org.springframework.security.web.server.ServerAuthenticationEntryPoint
import org.springframework.security.web.server.authorization.AuthorizationContext
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler
import org.springframework.security.web.server.util.matcher.NegatedServerWebExchangeMatcher
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers

@Configuration
@EnableWebFluxSecurity
@RequiredArgsConstructor
class SecurityConfig(
    private val jwtService: JwtService,
    private val rcServerAuthenticationEntryPoint: ServerAuthenticationEntryPoint,
    private val rcServerAccessDeniedHandler: ServerAccessDeniedHandler,
    private val rcReactiveAuthorizationManager: ReactiveAuthorizationManager<AuthorizationContext>,
) {

    val PUBLIC_ACCESS_PATH = arrayOf(
        "/auth/**",
        "/webjars/**",
        "/swagger-ui.html/**"
    )

    @Bean
    fun securityFilter(http: ServerHttpSecurity): SecurityWebFilterChain {
        http
            .logout(ServerHttpSecurity.LogoutSpec::disable)
            .csrf(ServerHttpSecurity.CsrfSpec::disable)
            .httpBasic(ServerHttpSecurity.HttpBasicSpec::disable)
            .formLogin(ServerHttpSecurity.FormLoginSpec::disable)

            // 하위 경로는 Filter 무시
            .securityMatcher(NegatedServerWebExchangeMatcher(ServerWebExchangeMatchers.pathMatchers(*PUBLIC_ACCESS_PATH)))

            .authorizeExchange { authorize ->
                authorize
                    .pathMatchers(*PUBLIC_ACCESS_PATH).permitAll()
                    .anyExchange().access(rcReactiveAuthorizationManager)
            }

            .addFilterBefore(RcJwtFilter(jwtService), SecurityWebFiltersOrder.AUTHENTICATION)

            .exceptionHandling {
                it.authenticationEntryPoint(rcServerAuthenticationEntryPoint)
                it.accessDeniedHandler(rcServerAccessDeniedHandler)
            }
        return http.build()
    }
}